requests_ntlm/requests_ntlm.py

```diff
@@ -180,3 +178,32 @@ def _auth_type_from_header(header):
         return 'Negotiate'
 
     return None
+
+def _get_server_cert(request_url):
+    """
+    Get the certificate at the request_url and return it as a SHA256 hash. Will check the endpoint if it
+    is a https site and use the default port 443 is it isn't explicitly specified in the URL. Used to send
+    in with NTLMv2 authentication for Channel Binding Tokens
+
+    :param request_url: The request url in the format https://endpoint:port/path
+    :return: SHA256 hash of the DER encoded certificate at the request_url or None if not a HTTPS endpoint
+    """
+    host_pattern = re.compile('(?i)^https://?(?P<host>[0-9a-z-_.]+)(:(?P<port>\d+))?')
+    match = host_pattern.match(request_url)
+
+    if match:
+        host = match.group('host')
+        port = match.group('port')
+        if not port:
+            port = 443
+        else:
+            port = int(port)
+
+        s = socket()
+        c = ssl.wrap_socket(s)
+        c.connect((host, port))
+        server_certificate = c.getpeercert(True)
+        hash_object = hashlib.sha256(server_certificate)
+        return hash_object.hexdigest().upper()
```

Comment:
The problems that limit most cases for getting the cert from requests do not apply in this case. There are two key issues:

1. Requests provides no public API for this. That's true, but in practice the socket always lives in basically one of two places: `response.raw._fp.fp._sock` (Python 2) or `response.raw._fp.fp.raw._sock` (Python 3).
2. The biggest problem with generally using that is that sometimes `response.raw._fp` is `None`, but that only happens when the connection has been terminated after the response has been totally read. That doesn't happen until *after* the response hook has fired, which means requests-ntlm can assume (or at least assert) that that's always there.

Now, we may want to tolerate non-urllib3-backends, but my suggestion is that for non-urllib3 backends we don't bother trying to get the peer certificate. That means that we can test to see if either of those names exists, and if they don't we can say that there is no peer cert. Otherwise, if either exists, we can grab the peer cert and be on our way.

Does that seem sensible?